### ISSUE TITLE
errors intermittently appear inside `if not TYPE_CHECKING` blocks

### ISSUE THREAD

**Issue by ItsDrike** [DetachHead/basedpyright#492]
BasedPyright incorrectly marks valid imports conditionally imported in `not TYPE_CHECKING` (or an else branch to `if TYPE_CHECKING`) as unknown, even though it is able to resolve them within the `TYPE_CHECKING` block.

```python
from typing import TYPE_CHECKING

if not TYPE_CHECKING:
    from functools import lru_cache  # basedpyright: "lru_cache" is unknown import symbol [reportAttributeAccessIssue]
```

Imports like this are a fairly common practice, though not in the way shown in the minimal repro example, instead projects usually define signatures in `if TYPE_CHECKING`, with the real import being in the `else` branch of the condition. Here's a quick example of that, to better understand why this is important:

```python
# lru_cache doesn't forward the call parameters in _lru_cache_wrapper type, only the return type,
# this fixes the issue, though it means losing the type info about the function being cached,
# that's annoying, but preserving the parameters is much more important to us, so this is the lesser
# evil from the two
if TYPE_CHECKING:
    from typing import TypeVar
    T = TypeVar("T")
    def lru_cache(func: T, /) -> T: ...
else:
    from functools import lru_cache
```

**Comment by DetachHead** [comment]
hmm i can't reproduce this. can you share your config, basedpyright version and python version? does it also happen on upstream pyright?

**Comment by ItsDrike** [comment]
Hmm, this is weird one, I'm using neovim and have basedpyright installed using Mason:![image](https://github.com/DetachHead/basedpyright/assets/20902250/f5c5bfbb-40e7-434a-9664-198b88cb5be7)

This is the output from `basedpyright --version`:
![image](https://github.com/DetachHead/basedpyright/assets/20902250/ab99f818-1fc4-4d23-b391-48bff8227897)

Interestingly enough, I do not see any errors reported when running `basedpyright` manually:
![image](https://github.com/DetachHead/basedpyright/assets/20902250/868697f7-7ed5-4706-b68b-42a4fe8a4992)

However, when running in my editor, I do get the error:
![image](https://github.com/DetachHead/basedpyright/assets/20902250/de627556-24ab-43cb-8829-5470c7ffb6ff)

I don't need to have any special settings set for `basedpyright` specified, I'm getting this even when just using the defaults.

This is the output from `:LspInfo`, so it's clearly using the mason-installed `basedpyright-langserver` binary
![image](https://github.com/DetachHead/basedpyright/assets/20902250/98079506-0ad1-4dd4-b7e8-6ed350ef02a6)

I have no idea what's going on here and I'm not even sure how to debug this further. Note that I have tried installing `basedpyright` into a venv and running neovim from inside of an activated environment. I did confirm that neovim was using the basedpyright version from the venv and yet I got the same error.

**Comment by ItsDrike** [comment]
When using regular `pyright`, I do not get this error in my editor:

![image](https://github.com/DetachHead/basedpyright/assets/20902250/686add1e-41a9-41f4-87fb-ec8cad7e9909)
![image](https://github.com/DetachHead/basedpyright/assets/20902250/eab9f0f0-2c29-4ccc-afa2-e7717c469fcf)

(Then again, regular pyright doesn't even check that code.)

**Comment by ItsDrike** [comment]
I have also tried running the langserver binary and checking if the diagnostic gets produced in an isolated setting, to do this, I used [`pylspclient`](https://github.com/yeger00/pylspclient).

<details>
<summary>(See code below:)</summary>

```python
from __future__ import annotations

import os
import subprocess
from pathlib import Path
from typing import Any

from pylspclient import JsonRpcEndpoint, LspClient, LspEndpoint
from pylspclient.lsp_pydantic_strcuts import LanguageIdentifier, TextDocumentItem

TEST_FILE = Path(__file__).parent / "example.py"
BASED_PYRIGHT_EXE = Path("~/.local/share/nvim/mason/bin/basedpyright-langserver").expanduser()


def main() -> None:
    process = subprocess.Popen(
        [BASED_PYRIGHT_EXE, "--stdio"],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )

    def callback(x: Any, /) -> None:
        print(f"Callback called with: {x!r}")

    json_rpc_endpoint = JsonRpcEndpoint(process.stdin, process.stdout)
    lsp_endpoint = LspEndpoint(
        json_rpc_endpoint,
        method_callbacks={},
        notify_callbacks={
            "textDocument/publishDiagnostics": callback,
            "window/logMessage": callback,
        },
    )
    client = LspClient(lsp_endpoint)

    client.initialize(
        processId=os.getpid(),
        rootPath=None,  # deprecated
        rootUri=None,  # Assume we're working in a single file (no workspace)
        initializationOptions=None,  # no idea what this even is
        capabilities={},  # we don't support any LSP capabilities
        trace=None,  # we don't want to log LSP messages
        workspaceFolders=None,  # we don't support workspace folders
    )

    # Tell the server we're ready
    client.initialized()

    # Tell the server we've just opened a file
    with TEST_FILE.open("r") as f:
        file_content = f.read()

    text_document = TextDocumentItem(
        uri=TEST_FILE.as_uri(),
        languageId=LanguageIdentifier.PYTHON,
        version=1,
        text=file_content,
    )
    client.didOpen(text_document)


if __name__ == "__main__":
    main()
```

</details>

<details>
<summary>Yet no diagnostics were reported at all</summary>

![image](https://github.com/DetachHead/basedpyright/assets/20902250/7de75bc9-8e8b-452a-b130-0d80afa10877)

</details>

I did also try reinstalling `basedpyright`, which did not help either.

**Comment by ItsDrike** [comment]
Okay, I figured out what was going on, my editor was configuring basedpyright's `typeCheckingMode` to `basic`, with that mode, I can consistently reproduce the issue.

```toml
[tool.basedpyright]
typeCheckingMode = "basic"
```

**Comment by ItsDrike** [comment]
I managed to put together a minimal reproducable example that demonstrates the issue with neovim. It uses `NVIM_APPNAME` to work in an isolated environment (so it won't affect your normal neovim setup) and the example just relies on using the lspconfig plugin. Note that `basedpyright-langserver`  needs to be in `$PATH`.

```bash
export NVIM_APPNAME="nvim_basedpython_test"
mkdir -p ~/.config/$NVIM_APPNAME/pack/nvim/start
git clone https://github.com/neovim/nvim-lspconfig ~/.config/$NVIM_APPNAME/pack/nvim/start/nvim-lspconfig
echo "require'lspconfig'.basedpyright.setup{}" > ~/.config/$NVIM_APPNAME/init.lua
cat <<EOF > ~/test.py
from typing import TYPE_CHECKING

if not TYPE_CHECKING:
    from functools import lru_cache
EOF
nvim ~/test.py  # test.py being the original problematic code
```

(Here's also a quick command to clear everything up afterwards)
```bash
rm -rf ~/.config/$NVIM_APPNAME ~/.local/{share,state}/$NVIM_APPNAME ~/test.py
```

**Comment by KotlinIsland** [comment]
Are other errors reported within the type checking blocks?
```py
if TYPE_CHECKING:
    1 + None
```

**Comment by DetachHead** [comment]
very strange... i can reproduce the issue in pycharm with the lsp4j plugin, but not vscode. the errors also seem to intermittently appear and disappear:

![](https://github.com/user-attachments/assets/cbb82993-d444-431c-8e70-6b1c862f0561)

possibly related: #69

**Comment by ItsDrike** [comment]
> @ItsDrike could you please check if #503 fixes it?
> 
> ```
> pip install git+https://github.com/DetachHead/basedpyright@be05a9d3f5094f921bd0ef91ce3f4a223b340915
> ```

Yes! Well, at least the unknown import symbol one.

> Are other errors reported within the type checking blocks?
> 
> ```python
> if not TYPE_CHECKING:
>     1 + None
> ```

Hmm, the `1 + None` example actually doesn't trigger an error (neither in the latest release nor in the suggested git version above). I'm not sure if this is the same issue, but it's definitely an issue.

![image](https://github.com/user-attachments/assets/8eecc6f8-e0c5-4948-9c05-2db2f1fb8a3e)
